File: src/main/kotlin/dev/ligature/Ligature.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package dev.ligature

data class Dataset(val name: String)

sealed class Object
sealed class Node: Object()
data class NamedNode(val name: String): Node()
data class AnonymousNode(val identifier: Long): Node()

sealed class Literal: Object()
data class LangLiteral(val value: String, val langTag: String): Literal()
data class StringLiteral(val value: String): Literal()
data class BooleanLiteral(val value: Boolean): Literal()
data class LongLiteral(val value: Long): Literal()
data class DoubleLiteral(val value: Double): Literal()

sealed class Range
data class LangLiteralRange(val start: LangLiteral, val stop: LangLiteral): Range()
data class StringLiteralRange(val start: StringLiteral, val stop: StringLiteral): Range()
data class LongLiteralRange(val start: LongLiteral, val stop: LongLiteral): Range()
data class DoubleLiteralRange(val start: DoubleLiteral, val stop: DoubleLiteral): Range()

data class Statement(val subject: Node, val predicate: NamedNode, val `object`: Object)
data class PersistedStatement(val dataset: Dataset, val statement: Statement, val context: AnonymousNode)

val a: NamedNode = NamedNode("http://www.w3.org/1999/02/22-rdf-syntax-ns#type")
fun validDataset(dataset: Dataset): Boolean =
  "[a-z_]+(/[a-z_]+)*".r.matches(dataset.name)

fun validNamedNode(node: NamedNode): Boolean =
  "[a-zA-Z_][^\\s()\\[\\]{}'\"`<>\\\\]*".r.matches(node.name)

fun validLangTag(langTag: String): Boolean =
  "[a-zA-Z]+(-[a-zA-Z0-9]+)*".r.matches(langTag)

interface Ligature {
  fun instance: Resource[IO, LigatureInstance]
}

interface LigatureInstance {
  fun read: Resource[IO, LigatureReadTx]
  fun write: Resource[IO, LigatureWriteTx]
}

interface LigatureReadTx {
  fun datasets: Stream[IO, Dataset]
  fun datasets(prefix: Dataset): Stream[IO, Dataset]
  fun datasets(from: Dataset, to: Dataset): Stream[IO, Dataset]
  fun allStatements(dataset: Dataset): Stream[IO, PersistedStatement]
  fun matchStatements(dataset: Dataset,
                      subject: Option[Node] = None,
                      predicate: Option[NamedNode] = None,
                      `object`: Option[Object] = None): Stream[IO, PersistedStatement]
  fun matchStatements(dataset: Dataset,
                      subject: Option[Node],
                      predicate: Option[NamedNode],
                      range: Range): Stream[IO, PersistedStatement]
  fun statementByContext(dataset: Dataset, context: AnonymousNode): IO[Option[PersistedStatement]]
}

interface LigatureWriteTx {
  fun createDataset(dataset: Dataset): IO[Dataset]
  fun deleteDataset(dataset: Dataset): IO[Dataset]
  fun newNode(dataset: Dataset): IO[AnonymousNode]
  fun addStatement(dataset: Dataset, statement: Statement): IO[PersistedStatement]
  fun removeStatement(dataset: Dataset, statement: Statement): IO[Statement]
  fun cancel(): Unit
}

File: src/main/scala/dev/ligature/Ligature.scala
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package dev.ligature

import cats.effect.Resource
import monix.eval.Task
import monix.reactive.Observable
import dev.ligature.iris.IRI

final case class LocalName(name: String)
final case class BlankNode(identifier: Long)
object DefaultGraph

sealed trait Literal
final case class LangLiteral(value: String, langTag: String) extends Literal
final case class StringLiteral(value: String) extends Literal
final case class BooleanLiteral(value: Boolean) extends Literal
final case class LongLiteral(value: Long) extends Literal
final case class DoubleLiteral(value: Double) extends Literal
final case class UnknownLiteral(value: String, `type`: IRI) extends Literal

sealed trait Range
final case class LangLiteralRange(start: LangLiteral, stop: LangLiteral) extends Range
final case class StringLiteralRange(start: StringLiteral, stop: StringLiteral) extends Range
final case class LongLiteralRange(start: LongLiteral, stop: LongLiteral) extends Range
final case class DoubleLiteralRange(start: DoubleLiteral, stop: DoubleLiteral) extends Range

type Subject = IRI | LocalName | BlankNode | DefaultGraph.type
type Graph = IRI | LocalName | BlankNode | DefaultGraph.type
type Object = Subject | Literal

final case class Statement(subject: Subject, predicate: IRI, `object`: Object)
final case class PersistedStatement(dataset: IRI, statement: Statement, graph: Graph)

object Ligature {
  val a: IRI = IRI("http://www.w3.org/1999/02/22-rdf-syntax-ns#type").getOrElse(???)
  def validLangTag(langTag: String): Boolean =
    "[a-zA-Z]+(-[a-zA-Z0-9]+)*".r.matches(langTag)
}

trait Ligature {
  def instance: Resource[Task, LigatureInstance]
}

trait LigatureInstance {
  def read: Resource[Task, LigatureReadTx]
  def write: Resource[Task, LigatureWriteTx]
}

trait LigatureReadTx {
  def datasets: Observable[IRI]
  def datasets(prefix: String): Observable[IRI]
  def datasets(from: String, to: String): Observable[IRI]
  def allStatements(dataset: IRI): Observable[PersistedStatement]
  def matchStatements(dataset: IRI,
                      subject: Option[Subject] = None,
                      predicate: Option[IRI] = None,
                      `object`: Option[Object] = None,
                      graph: Option[Graph] = None): Observable[PersistedStatement]
  def matchStatements(dataset: IRI,
                      subject: Option[Subject],
                      predicate: Option[IRI],
                      graph: Option[Graph],
                      range: Range): Observable[PersistedStatement]
  def statementByContext(dataset: IRI, context: BlankNode): Task[Option[PersistedStatement]]
}

trait LigatureWriteTx {
  def createDataset(dataset: IRI): Task[IRI]
  def deleteDataset(dataset: IRI): Task[IRI]
  def newNode(dataset: IRI): Task[BlankNode]
  def addStatement(dataset: IRI, statement: Statement, graph: Graph = DefaultGraph): Task[PersistedStatement]
  def removeStatement(dataset: IRI, statement: Statement, graph: Graph = DefaultGraph): Task[Statement]
  def cancel(): Unit
}
